src/lib/coerce.js

```diff
@@ -19,21 +19,70 @@ var nestedProperty = require('./nested_property');
 var counterRegex = require('./regex').counter;
 var modHalf = require('./mod').modHalf;
 var isArrayOrTypedArray = require('./array').isArrayOrTypedArray;
+var isTypedArraySpec = require('./array').isTypedArraySpec;
+var decodeTypedArraySpec = require('./array').decodeTypedArraySpec;
+var coerceTypedArraySpec = require('./array').coerceTypedArraySpec;
+
 
 exports.valObjectMeta = {
     data_array: {
         // You can use *dflt=[] to force said array to exist though.
         description: [
             'An {array} of data.',
-            'The value MUST be an {array}, or we ignore it.',
-            'Note that typed arrays (e.g. Float32Array) are supported.'
+            'The value could be an {array}',
+            'noting that typed arrays (e.g. Float32Array) are also supported.',
+            'It could also be an object in the form of',
+            'v: {, dtype: \'float32\', bvals: [/* ... */]}, shape: [dim0 (, dim1, (dim3))]',
+            'otherwise, it would be ignored.'
         ].join(' '),
         requiredOpts: [],
         otherOpts: ['dflt'],
         coerceFunction: function(v, propOut, dflt) {
-            // TODO maybe `v: {type: 'float32', vals: [/* ... */]}` also
-            if(isArrayOrTypedArray(v)) propOut.set(v);
-            else if(dflt !== undefined) propOut.set(dflt);
+            var wasSet;
+            if(isArrayOrTypedArray(v)) {
+                propOut.set(v);
+                wasSet = true;
+            } else if(isTypedArraySpec(v)) {
+                // Copy and coerce spec
+                v = coerceTypedArraySpec(v);
+
+                // See if caching location is available
+                var uid = propOut.obj.uid;
+                var module = propOut.obj._module;
```

Comment:
Why are we attaching this to the `_module`? Why not just one cache, stored here, keyed to the `uid`?

Also I don't think we'll always have a `uid` (or `_module`) - that depends on `propOut` being based at the root of the trace. For example in `splom`, `parcoords`, and `parcats` the data mostly lives in `dimensions[i].values` but we're just calling [`coerce('values')`](https://github.com/plotly/plotly.js/blob/fa781edd26bc8f9406e170310ec15a47fd90b8bb/src/traces/splom/defaults.js#L62) from each dimension object. I'm not sure how to handle this... only things I can think of are setting some global var early in each trace's `supplyDefaults`, and override it as necessary when calling `coerce` from a sub-object... or passing in another arg to `coerce` from these sub-objects, and feeding that through to the `coerceFunction`.

Reply:
Actually I'm a bit worried about caching by `uid` even, that it will be a memory leak over time, particularly if traces (and plots) get added and removed so we have uids in the cache that don't exist on the page. Can we find a way to store a cache in each `_fullData` trace, make sure it's relinked early in the `supplyDefaults` process so it's available here, and pass that through if we're in a sub-container as mentioned above?

Reply:
Yeah, those are all good points. To start with I was just looking for somewhere persistent that was accessible from inside coerce.  The `_fullData` approach does make sense for managing the cache lifetimes here.